SPEC.rdoc

```diff
@@ -1,103 +1,165 @@
+
+==== <tt>rack.protocol</tt>
+
+An optional +Array+ of +String+ values, containing the protocols advertised by the client in the <tt>upgrade</tt> header (HTTP/1) or the <tt>:protocol</tt> pseudo-header (HTTP/2+).
+
+==== <tt>rack.session</tt>
+
+An optional +Hash+-like interface for storing request session data. The store must implement:
+* <tt>store(key, value)</tt> (aliased as <tt>[]=</tt>) to set a value for a key,
+* <tt>fetch(key, default = nil)</tt> (aliased as <tt>[]</tt>) to retrieve a value for a key,
+* <tt>delete(key)</tt> to delete a key,
+* <tt>clear</tt> to clear the session,
+* <tt>to_hash</tt> (optional) to retrieve the session as a Hash.
+
+==== <tt>rack.logger</tt>
+
+An optional +Logger+-like interface for logging messages. The logger must implement:
+* <tt>info(message, &block)</tt>,
+* <tt>debug(message, &block)</tt>,
+* <tt>warn(message, &block)</tt>,
+* <tt>error(message, &block)</tt>,
+* <tt>fatal(message, &block)</tt>.
+
+==== <tt>rack.multipart.buffer_size</tt>
+
+An optional +Integer+ hint to the multipart parser as to what chunk size to use for reads and writes.
+
+==== <tt>rack.multipart.tempfile_factory</tt>
+
+An optional object for constructing temporary files for multipart form data. The factory must implement:
+* <tt>call(filename, content_type)</tt> to create a temporary file for a multipart form field.
+The factory must return an +IO+-like object that responds to <tt><<</tt> and optionally <tt>rewind</tt>.
+
+==== <tt>rack.hijack?</tt>
+
+If present and true, indicates that the server supports partial hijacking. See the section below on hijacking for more information.
+
+==== <tt>rack.hijack</tt>
+
+If present, an object responding to +call+ that is used to perform a full hijack. See the section below on hijacking for more information.
+
+==== <tt>rack.early_hints</tt>
+
+If present, an object responding to +call+ that is used to send early hints. See the section below on early hints for more information.
+
+==== <tt>rack.input</tt>
+
+If present, the input stream. See the section below on the input stream for more information.
+
+==== <tt>rack.errors</tt>
+
+The error stream. See the section below on the error stream for more information.
+
+==== <tt>rack.response_finished</tt>
+
+If present, an array of callables that will be run by the server after the response has been processed. The callables are called with <tt>environment, status, headers, error</tt> arguments and should not raise any exceptions. The callables would typically be called after sending the response to the client, but it could also be called if an error occurs while generating the response or sending the response (in that case, the +error+ argument will be a kind of +Exception+). The callables will be invoked in reverse order.
```

Comment:
```suggestion
If present, an array of callables that will be run by the server after the response has been processed. The callables are called with <tt>environment, status, headers, error</tt> arguments and should not raise any exceptions. The callables would typically be called after sending the response to the client, but it could also be called if an error occurs while generating the response or sending the response (in that case, the +error+ argument will be a kind of +Exception+). The callables will be called in reverse order.
```

Reply:
The original documentation used the word invoked and I am okay with not changing it as the word invoked is used extensively throughout the document. In addition, I find "The callable will be called" clumsy. If you want to change this, I suggest you follow up with a PR and make it consistent everywhere.